{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "root = '../Projekat_1/'\n",
    "sample_folders = ['GSM2560245', 'GSM2560246', 'GSM2560247', 'GSM2560248',\n",
    "                 'GSM2560249', 'GSM3087619', 'GSM3087622', 'GSM3087624',\n",
    "                 'GSM3087626', 'GSM3169075', 'GSM3374613', 'GSM3374614',\n",
    "                 'GSM3478791', 'GSM3478792', 'GSM3892570', 'GSM3892571', \n",
    "                 'GSM3892572', 'GSM3892573', 'GSM3892574', 'GSM3892575', \n",
    "                 'GSM3892576']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>INDEX</th>\n",
       "      <th>SERIES</th>\n",
       "      <th>ID:</th>\n",
       "      <th>ACCESSION</th>\n",
       "      <th>No SAMPLES SER</th>\n",
       "      <th>GENOME</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>No SAMPLES</th>\n",
       "      <th>...</th>\n",
       "      <th>AUTHOR DESCRIPTION</th>\n",
       "      <th>DESCRIPTION</th>\n",
       "      <th>Comment:</th>\n",
       "      <th>Date</th>\n",
       "      <th>Unnamed: 18</th>\n",
       "      <th>Unnamed: 19</th>\n",
       "      <th>Unnamed: 20</th>\n",
       "      <th>Unnamed: 21</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "      <th>Unnamed: 23</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1462</td>\n",
       "      <td>218</td>\n",
       "      <td>200130318</td>\n",
       "      <td>GSE130318</td>\n",
       "      <td>2</td>\n",
       "      <td>Ensembl_GRCh38.p12_rel94</td>\n",
       "      <td>GSM3735993</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>(Submitter supplied) Single human primary pros...</td>\n",
       "      <td>Monolayer; Primary Prostate Epithelial Cells; ...</td>\n",
       "      <td>Ensembl_GRCh38.p12_rel94, probes are 58735, no...</td>\n",
       "      <td>11-Jul-19</td>\n",
       "      <td>OK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1463</td>\n",
       "      <td>218</td>\n",
       "      <td>200130318</td>\n",
       "      <td>GSE130318</td>\n",
       "      <td>2</td>\n",
       "      <td>Ensembl_GRCh38.p12_rel94</td>\n",
       "      <td>GSM3735994</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>(Submitter supplied) Single human primary pros...</td>\n",
       "      <td>Organoid; Primary Prostate Epithelial Cells; H...</td>\n",
       "      <td>Ensembl_GRCh38.p12_rel94, probes are 58735, no...</td>\n",
       "      <td>11-Jul-19</td>\n",
       "      <td>OK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "      <td>8</td>\n",
       "      <td>200102596</td>\n",
       "      <td>GSE102596</td>\n",
       "      <td>1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>GSM2741551</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>(Submitter supplied) To characterize the cellu...</td>\n",
       "      <td>human embryonic kidney cortex cells week 16</td>\n",
       "      <td></td>\n",
       "      <td>1-Jan-19</td>\n",
       "      <td>OK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>323</td>\n",
       "      <td>57</td>\n",
       "      <td>200112570</td>\n",
       "      <td>GSE112570</td>\n",
       "      <td>2</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>GSM3073088</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>(Submitter supplied) To characterize the cellu...</td>\n",
       "      <td>human week 17 kidney cortex single cell RNA-Se...</td>\n",
       "      <td></td>\n",
       "      <td>13-Apr-18</td>\n",
       "      <td>OK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>324</td>\n",
       "      <td>57</td>\n",
       "      <td>200112570</td>\n",
       "      <td>GSE112570</td>\n",
       "      <td>2</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>GSM3073089</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>(Submitter supplied) To characterize the cellu...</td>\n",
       "      <td>human week 17 kidney cortex single cell RNA-Se...</td>\n",
       "      <td></td>\n",
       "      <td>13-Apr-18</td>\n",
       "      <td>OK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  INDEX  SERIES        ID:  ACCESSION  No SAMPLES SER  \\\n",
       "0           1   1462     218  200130318  GSE130318               2   \n",
       "1           2   1463     218  200130318  GSE130318               2   \n",
       "2           3     50       8  200102596  GSE102596               1   \n",
       "3           4    323      57  200112570  GSE112570               2   \n",
       "4           5    324      57  200112570  GSE112570               2   \n",
       "\n",
       "                     GENOME      SAMPLE      ORGANISM  No SAMPLES  ...  \\\n",
       "0  Ensembl_GRCh38.p12_rel94  GSM3735993  Homo sapiens           1  ...   \n",
       "1  Ensembl_GRCh38.p12_rel94  GSM3735994  Homo sapiens           1  ...   \n",
       "2                    GRCh37  GSM2741551  Homo sapiens           1  ...   \n",
       "3                    GRCh37  GSM3073088  Homo sapiens           1  ...   \n",
       "4                    GRCh37  GSM3073089  Homo sapiens           1  ...   \n",
       "\n",
       "                                  AUTHOR DESCRIPTION  \\\n",
       "0  (Submitter supplied) Single human primary pros...   \n",
       "1  (Submitter supplied) Single human primary pros...   \n",
       "2  (Submitter supplied) To characterize the cellu...   \n",
       "3  (Submitter supplied) To characterize the cellu...   \n",
       "4  (Submitter supplied) To characterize the cellu...   \n",
       "\n",
       "                                         DESCRIPTION  \\\n",
       "0  Monolayer; Primary Prostate Epithelial Cells; ...   \n",
       "1  Organoid; Primary Prostate Epithelial Cells; H...   \n",
       "2       human embryonic kidney cortex cells week 16    \n",
       "3  human week 17 kidney cortex single cell RNA-Se...   \n",
       "4  human week 17 kidney cortex single cell RNA-Se...   \n",
       "\n",
       "                                            Comment:       Date Unnamed: 18  \\\n",
       "0  Ensembl_GRCh38.p12_rel94, probes are 58735, no...  11-Jul-19          OK   \n",
       "1  Ensembl_GRCh38.p12_rel94, probes are 58735, no...  11-Jul-19          OK   \n",
       "2                                                      1-Jan-19          OK   \n",
       "3                                                     13-Apr-18          OK   \n",
       "4                                                     13-Apr-18          OK   \n",
       "\n",
       "  Unnamed: 19 Unnamed: 20 Unnamed: 21 Unnamed: 22 Unnamed: 23  \n",
       "0         NaN         NaN         NaN         NaN         NaN  \n",
       "1         NaN         NaN         NaN         NaN         NaN  \n",
       "2         NaN         NaN         NaN         NaN         NaN  \n",
       "3         NaN         NaN         NaN         NaN         NaN  \n",
       "4         NaN         NaN         NaN         NaN         NaN  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uzorci_data = pd.read_csv(os.path.join(root, 'uzorci.seminarski.csv'))\n",
    "uzorci_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG_ID</th>\n",
       "      <th>hg19</th>\n",
       "      <th>hg37</th>\n",
       "      <th>hg38</th>\n",
       "      <th>Ensembl_GRCh38.p12_rel94</th>\n",
       "      <th>GSM3717979</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000181638</td>\n",
       "      <td>hg19_ZFP41</td>\n",
       "      <td>grch37_ZFP41</td>\n",
       "      <td>grch38_ZFP41</td>\n",
       "      <td>#ZFP41</td>\n",
       "      <td>#ZFP41</td>\n",
       "      <td>in all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000111875</td>\n",
       "      <td>hg19_ASF1A</td>\n",
       "      <td>grch37_ASF1A</td>\n",
       "      <td>grch38_ASF1A</td>\n",
       "      <td>#ASF1A</td>\n",
       "      <td>#ASF1A</td>\n",
       "      <td>in all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000176142</td>\n",
       "      <td>hg19_TMEM39A</td>\n",
       "      <td>grch37_TMEM39A</td>\n",
       "      <td>grch38_TMEM39A</td>\n",
       "      <td>#TMEM39A</td>\n",
       "      <td>#TMEM39A</td>\n",
       "      <td>in all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000177186</td>\n",
       "      <td>hg19_OR2M7</td>\n",
       "      <td>grch37_OR2M7</td>\n",
       "      <td>grch38_OR2M7</td>\n",
       "      <td>#OR2M7</td>\n",
       "      <td>#OR2M7</td>\n",
       "      <td>in all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000135624</td>\n",
       "      <td>hg19_CCT7</td>\n",
       "      <td>grch37_CCT7</td>\n",
       "      <td>grch38_CCT7</td>\n",
       "      <td>#CCT7</td>\n",
       "      <td>#CCT7</td>\n",
       "      <td>in all</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ENSG_ID          hg19            hg37            hg38  \\\n",
       "0  ENSG00000181638    hg19_ZFP41    grch37_ZFP41    grch38_ZFP41   \n",
       "1  ENSG00000111875    hg19_ASF1A    grch37_ASF1A    grch38_ASF1A   \n",
       "2  ENSG00000176142  hg19_TMEM39A  grch37_TMEM39A  grch38_TMEM39A   \n",
       "3  ENSG00000177186    hg19_OR2M7    grch37_OR2M7    grch38_OR2M7   \n",
       "4  ENSG00000135624     hg19_CCT7     grch37_CCT7     grch38_CCT7   \n",
       "\n",
       "  Ensembl_GRCh38.p12_rel94 GSM3717979 Unnamed: 6  \n",
       "0                   #ZFP41     #ZFP41     in all  \n",
       "1                   #ASF1A     #ASF1A     in all  \n",
       "2                 #TMEM39A   #TMEM39A     in all  \n",
       "3                   #OR2M7     #OR2M7     in all  \n",
       "4                    #CCT7      #CCT7     in all  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_hlist_data = pd.read_csv(os.path.join(root, 'common_human_list.csv'))\n",
    "common_hlist_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sample_genome(sample_id):\n",
    "    sample_filter = uzorci_data['SAMPLE'] == sample_id\n",
    "    uzorci_seminarski_data_sample = uzorci_data[sample_filter]\n",
    "    genome = uzorci_seminarski_data_sample['GENOME'].values[0]\n",
    "    print(f'{sample_id} <=> {genome}')\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 21/21 [00:00<00:00, 896.99it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM2560245 <=> hg19\n",
      "GSM2560246 <=> hg19\n",
      "GSM2560247 <=> hg19\n",
      "GSM2560248 <=> hg19\n",
      "GSM2560249 <=> hg19\n",
      "GSM3087619 <=> GRCh38\n",
      "GSM3087622 <=> GRCh38\n",
      "GSM3087624 <=> GRCh38\n",
      "GSM3087626 <=> GRCh38\n",
      "GSM3169075 <=> GRCh38 version 90\n",
      "GSM3374613 <=> hg38\n",
      "GSM3374614 <=> hg38\n",
      "GSM3478791 <=> GRCh38\n",
      "GSM3478792 <=> GRCh38\n",
      "GSM3892570 <=> GRCh38\n",
      "GSM3892571 <=> GRCh38\n",
      "GSM3892572 <=> GRCh38\n",
      "GSM3892573 <=> GRCh38\n",
      "GSM3892574 <=> GRCh38\n",
      "GSM3892575 <=> GRCh38\n",
      "GSM3892576 <=> GRCh38\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "sample_genomes = {}\n",
    "\n",
    "for i in tqdm(range(len(sample_folders))):\n",
    "    sample_genomes[sample_folders[i]] = get_sample_genome(sample_folders[i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_filename(folder):\n",
    "    path = os.path.join(root, folder)\n",
    "    CSV_Files = [file for file in os.listdir(path) if file.endswith('.csv')]\n",
    "    filename = CSV_Files[0]\n",
    "    return filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_common_hlist_ENSG(sample_genome):\n",
    "    common_hlist_ENSG = common_hlist_data.loc[:, ['ENSG_ID', sample_genome]]\n",
    "\n",
    "    return common_hlist_ENSG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(sample, dropna=True):\n",
    "    filename = get_filename(sample)\n",
    "    path = os.path.join(root, sample)\n",
    "    sample_data = pd.read_csv(os.path.join(path, filename))\n",
    "    common_hlist_ENSG = get_common_hlist_ENSG(sample_genomes[sample])\n",
    "    \n",
    "    sample_data_filtered_ENSG = sample_data.set_index('Index').join(common_hlist_ENSG.set_index('ENSG_ID'), how='left')\n",
    "#     print(f'Filtered shape: {sample_data_filtered_ENSG.shape}')\n",
    "    \n",
    "    if dropna:\n",
    "        sample_data = sample_data_filtered_ENSG.dropna()\n",
    "    else:\n",
    "        sample_data = sample_data_filtered_ENSG\n",
    "    \n",
    "    filtered_columns = sample_data.columns\n",
    "    num_cols = sample_data.shape[1]\n",
    "    \n",
    "    new_columns = [str(sample)[3:]+'_'+str(i) for i in range(1, num_cols)] + ['gene']\n",
    "    sample_data.columns = new_columns\n",
    "    \n",
    "    return sample_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample = read_data(sample_folders[0])\n",
    "# sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 14%|█▍        | 3/21 [01:29<08:21, 27.88s/it]"
     ]
    }
   ],
   "source": [
    "samples = []\n",
    "for i in tqdm(range(len(sample_folders))):\n",
    "    samples.append(read_data(sample_folders[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
